api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -543,6 +565,148 @@ public int hashCode() {
     }
   }
 
+  public static class GeometryType extends PrimitiveType {
+
+    private final String crs;
+
+    private GeometryType(String crs) {
+      if (crs != null) {
+        Preconditions.checkArgument(!crs.isEmpty(), "Invalid CRS: (empty string)");
+        Preconditions.checkArgument(
+            crs.trim().equals(crs), "CRS must not have leading or trailing spaces: '%s'", crs);
+        this.crs = crs;
+      } else {
+        this.crs = null;
+      }
+    }
+
+    private GeometryType() {
+      crs = null;
+    }
+
+    public static GeometryType crs84() {
+      return new GeometryType();
+    }
+
+    public static GeometryType of(String crs) {
+      return new GeometryType(crs);
+    }
+
+    @Override
+    public TypeID typeId() {
+      return TypeID.GEOMETRY;
+    }
+
+    public String crs() {
+      return crs;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      } else if (!(o instanceof GeometryType)) {
+        return false;
+      }
+
+      GeometryType that = (GeometryType) o;
+      return Objects.equals(crs, that.crs);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(GeometryType.class, crs);
+    }
+
+    @Override
+    public String toString() {
+      if (crs == null) {
+        return "geometry";
+      }
+
+      return String.format("geometry(%s)", crs);
+    }
+  }
+
+  public static class GeographyType extends PrimitiveType {
+
+    public static final String DEFAULT_CRS = "OGC:CRS84";
+
+    private final String crs;
+    private final EdgeAlgorithm algorithm;
+
+    private GeographyType(String crs, EdgeAlgorithm algorithm) {
+      if (crs != null) {
+        Preconditions.checkArgument(!crs.isEmpty(), "Invalid CRS: (empty string)");
+        Preconditions.checkArgument(
+            crs.trim().equals(crs), "CRS must not have leading or trailing spaces: '%s'", crs);
+        this.crs = crs;
+      } else {
+        this.crs = null;
+      }
+
+      this.algorithm = algorithm;
+    }
+
+    private GeographyType() {
+      this.crs = null;
+      this.algorithm = null;
+    }
+
+    public static GeographyType crs84() {
+      return new GeographyType();
+    }
+
+    public static GeographyType forCRS(String crs) {
```

Comment:
I think I disagree with @szehon-ho for the name here. I think that `of` is fine because the other version of this method accepts a CRS string and `EdgeAlgorithm`. I'm mainly motivated here by making this method match what we use for `geometry`.

Reply:
Sounds good

Reply:
Renamed it back to `of`.